File: src/flask/app/api/routes/integrations.py
"""Routes for third-party integration management."""

from flask import Blueprint, jsonify, request

from app.api.deps import get_current_user, get_db, require_scope
from app.models import IntegrationConfig
from app.repositories.integration_repo import IntegrationConfigRepository
from app.requests.integration_request import (
    CredentialUpdate,
    IntegrationCreate,
    IntegrationUpdate,
    TestConnectionRequest,
)
from app.responses.integration_response import (
    IntegrationConfigPublic,
    IntegrationConfigPublicWithCreds,
    IntegrationsPublic,
    TestConnectionResponse,
)
from app.services.integration_service import KNOWN_INTEGRATIONS, IntegrationService

router = Blueprint("integrations", __name__, url_prefix="/integrations")


def _get_service(session):
    return IntegrationService(IntegrationConfigRepository(session))


def _mask_credentials(creds):
    if not creds:
        return {}
    return {
        field: f"****{value[-4:]}" if len(value) > 4 else "****" for field, value in creds.items()
    }


@router.route("/", methods=["GET"])
@require_scope("integrations:admin")
def list_integrations(skip: int = 0, limit: int = 100):
    session = get_db()
    _ = get_current_user()
    service = _get_service(session)
    items, total = service.get_all(skip, limit)
    integrations = []
    for item in items:
        try:
            is_valid = IntegrationConfigPublic.model_validate(item.model_dump())
            if is_valid:
                integrations.append(is_valid)
        except Exception as error:
            print(error)
    return jsonify(IntegrationsPublic(data=integrations, count=total).model_dump())


@router.route("/status")
def get_integrations_status():
    session = get_db()
    from sqlalchemy import select

    statement = select(IntegrationConfig)
    result = session.execute(statement)
    integrations = list(result.scalars().all())
    return jsonify({i.type: {"enabled": i.enabled, "status": i.status} for i in integrations})


@router.route("/<integration_id>", methods=["GET"])
@require_scope("integrations:admin")
def get_integration(integration_id: str):
    session = get_db()
    _ = get_current_user()
    service = _get_service(session)
    integration = service.get_by_id(integration_id)
    if not integration:
        return jsonify({"detail": "Integration not found"}), 404
    creds = service.get_credentials(integration)
    return jsonify(
        IntegrationConfigPublicWithCreds(
            **IntegrationConfigPublic.model_validate(integration.model_dump()).model_dump(),
            credential_fields=_mask_credentials(creds),
        ).model_dump()
    )


@router.route("/", methods=["POST"])
@require_scope("integrations:admin")
def create_integration():
    session = get_db()
    _ = get_current_user()
    data = request.get_json()
    integration_in = IntegrationCreate(**data)
    service = _get_service(session)
    existing = service.get_by_type(integration_in.type)
    if existing:
        return jsonify({"detail": f"Integration type '{integration_in.type}' already exists"}), 409
    meta = KNOWN_INTEGRATIONS.get(integration_in.type, {})
    display_name = integration_in.display_name or meta.get("display_name", integration_in.type)
    icon = integration_in.icon or meta.get("icon", "Plug")
    integration = service.create(
        type=integration_in.type,
        display_name=display_name,
        icon=icon,
        enabled=integration_in.enabled,
        config_json=integration_in.config_json,
        credentials=integration_in.credentials,
    )
    creds = service.get_credentials(integration)
    return (
        jsonify(
            IntegrationConfigPublicWithCreds(
                **IntegrationConfigPublic.model_validate(integration.model_dump()).model_dump(),
                credential_fields=_mask_credentials(creds),
            ).model_dump()
        ),
        201,
    )


@router.route("/<integration_id>", methods=["PUT"])
@require_scope("integrations:admin")
def update_integration(integration_id: str):
    session = get_db()
    _ = get_current_user()
    data = request.get_json()
    integration_in = IntegrationUpdate(**data)
    service = _get_service(session)
    integration = service.get_by_id(integration_id)
    if not integration:
        return jsonify({"detail": "Integration not found"}), 404
    update_data = integration_in.model_dump(exclude_unset=True)
    updated = service.update(integration, update_data)
    creds = service.get_credentials(updated)
    return jsonify(
        IntegrationConfigPublicWithCreds(
            **IntegrationConfigPublic.model_validate(updated.model_dump()).model_dump(),
            credential_fields=_mask_credentials(creds),
        ).model_dump()
    )


@router.route("/<integration_id>/credentials", methods=["PATCH"])
@require_scope("integrations:admin")
def update_credentials(integration_id: str):
    session = get_db()
    _ = get_current_user()
    data = request.get_json()
    cred_in = CredentialUpdate(**data)
    service = _get_service(session)
    integration = service.get_by_id(integration_id)
    if not integration:
        return jsonify({"detail": "Integration not found"}), 404
    updated = service.update_credentials(integration, cred_in.credentials)
    creds = service.get_credentials(updated)
    return jsonify(
        IntegrationConfigPublicWithCreds(
            **IntegrationConfigPublic.model_validate(updated.model_dump()).model_dump(),
            credential_fields=_mask_credentials(creds),
        ).model_dump()
    )


@router.route("/<integration_id>", methods=["DELETE"])
@require_scope("integrations:admin")
def delete_integration(integration_id: str):
    session = get_db()
    _ = get_current_user()
    service = _get_service(session)
    integration = service.get_by_id(integration_id)
    if not integration:
        return jsonify({"detail": "Integration not found"}), 404
    service.delete(integration)
    return jsonify({"message": "Integration deleted"})


@router.route("/test-connection", methods=["POST"])
@require_scope("integrations:admin")
def test_connection():
    session = get_db()
    _ = get_current_user()
    data = request.get_json()
    test_in = TestConnectionRequest(**data)
    service = _get_service(session)
    config = None
    if test_in.config_json:
        import json

        try:
            config = json.loads(test_in.config_json)
        except json.JSONDecodeError:
            return jsonify({"detail": "Invalid config_json"}), 400
    result = service.test_connection(test_in.type, test_in.credentials, config)
    return jsonify(TestConnectionResponse(**result).model_dump())


@router.route("/sync-status/<integration_id>", methods=["POST"])
@require_scope("integrations:admin")
def sync_status(integration_id: str):
    session = get_db()
    _ = get_current_user()
    service = _get_service(session)
    integration = service.get_by_id(integration_id)
    if not integration:
        return jsonify({"detail": "Integration not found"}), 404
    updated = service.sync_status(integration, "connected")
    return jsonify(IntegrationConfigPublic.model_validate(updated.model_dump()).model_dump())


@router.route("/pre-seed", methods=["POST"])
@require_scope("integrations:admin")
def pre_seed_integrations():
    session = get_db()
    _ = get_current_user()
    service = _get_service(session)
    created = []
    for type_id, meta in KNOWN_INTEGRATIONS.items():
        existing = service.get_by_type(type_id)
        if not existing:
            created.append(
                service.create(
                    type=type_id,
                    display_name=meta["display_name"],
                    icon=meta["icon"],
                    enabled=False,
                    config_json=None,
                    credentials={},
                )
            )
    items, total = service.get_all()
    return jsonify(
        IntegrationsPublic(
            data=[IntegrationConfigPublic.model_validate(i.model_dump()) for i in items],
            count=total,
        ).model_dump()
    )

File: src/fe/lib/api/integrations.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "https://localhost:8000/";
const API_V1 = "api/v1";

import { fetchWithAuth } from "./auth";

// --- Integration types ---

export interface IntegrationConfig {
  id: string;
  type: string;
  display_name: string;
  icon: string;
  enabled: boolean;
  status: string;
  last_synced_at: string | null;
  config_json: string | null;
  created_on: string;
  updated_on: string | null;
}

export interface IntegrationCredentialField {
  type: string;
  fields: Record<string, string>;
}

export interface IntegrationWithCreds extends IntegrationConfig {
  credential_fields: Record<string, string>;
}

export interface IntegrationsResponse {
  data: IntegrationConfig[];
  count: number;
}

export interface TestConnectionResult {
  success: boolean;
  status: string;
  message: string;
}

// --- Integration API functions ---

export async function fetchIntegrationsStatus(): Promise<Record<string, { enabled: boolean; status: string }>> {
  const res = await fetch(`${API_BASE}${API_V1}/integrations/status`);
  if (!res.ok) {
    throw new Error("Failed to fetch integration status");
  }
  return res.json();
}

export async function fetchIntegrations(): Promise<IntegrationsResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/`);
  if (!res.ok) {
    throw new Error("Failed to fetch integrations");
  }
  return res.json();
}

export async function fetchIntegration(id: string): Promise<IntegrationWithCreds> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/${id}`);
  if (!res.ok) {
    throw new Error("Failed to fetch integration");
  }
  return res.json();
}

export async function createIntegration(data: {
  type: string;
  display_name: string;
  icon: string;
  enabled: boolean;
  config_json?: string | null;
  credentials: Record<string, string>;
}): Promise<IntegrationWithCreds> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to create integration");
  }
  return res.json();
}

export async function updateIntegration(id: string, data: Partial<{
  display_name: string;
  icon: string;
  enabled: boolean;
  status: string;
  config_json: string | null;
}>): Promise<IntegrationWithCreds> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/${id}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to update integration");
  }
  return res.json();
}

export async function updateIntegrationCredentials(
  id: string,
  credentials: Record<string, string>,
): Promise<IntegrationWithCreds> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/${id}/credentials`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ credentials }),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to update credentials");
  }
  return res.json();
}

export async function deleteIntegration(id: string): Promise<void> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/${id}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to delete integration");
  }
}

export async function testIntegrationConnection(data: {
  type: string;
  credentials: Record<string, string>;
  config_json?: string | null;
}): Promise<TestConnectionResult> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/test-connection`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Test connection failed");
  }
  return res.json();
}

export async function preSeedIntegrations(): Promise<IntegrationsResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/integrations/pre-seed`, {
    method: "POST",
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to pre-seed integrations");
  }
  return res.json();
}
